test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/DiscoverTests.cs

```diff
@@ -169,6 +171,42 @@ public void DiscoverTestsUsingSourceNavigation(RunnerInfo runnerInfo)
             }
         }
 
+        [TestMethod]
+        [NetFullTargetFrameworkDataSource]
+        [NetCoreTargetFrameworkDataSource]
+        public void CancelTestDiscovery(RunnerInfo runnerInfo)
+        {
+            // Setup
+            var testAssemblies = new List<string>
+                                     {
+                                         this.GetAssetFullPath("DiscoveryTestProject.dll"),
+                                         this.GetAssetFullPath("SimpleTestProject.dll"),
+                                         this.GetAssetFullPath("SimpleTestProject2.dll")
+                                     };
+
+            SetTestEnvironment(this.testEnvironment, runnerInfo);
+            this.Setup();
+
+            var discoveredTests = new List<TestCase>();
+            var discoveryEvents = new Mock<ITestDiscoveryEventsHandler>();
+            discoveryEvents.Setup((events) => events.HandleDiscoveredTests(It.IsAny<IEnumerable<TestCase>>())).Callback
+                ((IEnumerable<TestCase> testcases) => { discoveredTests.AddRange(testcases); });
+
+            // Act
+            var discoveryTask = Task.Run(() =>
+            {
+                this.vstestConsoleWrapper.DiscoverTests(testAssemblies, this.GetDefaultRunSettings(), discoveryEvents.Object);
+            });
+
+            Task.Delay(2000).Wait();
```

Comment:
We should not have this delay. This is possibly one of the scenarios. When the user immediately cancels.

In case of run, I recall a scenario where there was a race between the actual run request and cancel request.

Reply:
The delay is for the discovertest to actually start. It isn't an async operation. It would be good to cover the race condition through UTs and not through acceptance test as it is never going to validate the scenario in a deterministic